Job using "Direct Hadoop connection" cluster fails to run on Spark

I made a "Direct Hadoop connection" cluster with the name "Bigdatavm - direct", but when submitting to ```Run on Hadoop cluster``` it fails like this:

```
INFO  09:00:48.281 [Thread-24] ApplicationDriver - 16/03/30 09:00:48 INFO Client: 
INFO  09:00:48.281 [Thread-24] ApplicationDriver - 	 client token: N/A
INFO  09:00:48.281 [Thread-24] ApplicationDriver - 	 diagnostics: User class threw exception: java.lang.IllegalArgumentException: Illegal character in authority at index 7: hdfs://{Bigdatavm - direct}/weblog/weblog_2016_03_28.txt
INFO  09:00:48.281 [Thread-24] ApplicationDriver - 	 ApplicationMaster host: 192.168.33.10
INFO  09:00:48.281 [Thread-24] ApplicationDriver - 	 ApplicationMaster RPC port: 0
INFO  09:00:48.281 [Thread-24] ApplicationDriver - 	 queue: default
INFO  09:00:48.281 [Thread-24] ApplicationDriver - 	 start time: 1459321200642
INFO  09:00:48.282 [Thread-24] ApplicationDriver - 	 final status: FAILED
INFO  09:00:48.282 [Thread-24] ApplicationDriver - 	 tracking URL: http://bigdatavm:8088/proxy/application_1459321040367_0001/
INFO  09:00:48.282 [Thread-24] ApplicationDriver - 	 user: kaspers
INFO  09:00:48.303 [Thread-24] ApplicationDriver - Exception in thread "main" org.apache.spark.SparkException: Application application_1459321040367_0001 finished with failed status
INFO  09:00:48.303 [Thread-24] ApplicationDriver - 	at org.apache.spark.deploy.yarn.Client.run(Client.scala:920)
INFO  09:00:48.303 [Thread-24] ApplicationDriver - 	at org.apache.spark.deploy.yarn.Client$.main(Client.scala:966)
INFO  09:00:48.303 [Thread-24] ApplicationDriver - 	at org.apache.spark.deploy.yarn.Client.main(Client.scala)
INFO  09:00:48.303 [Thread-24] ApplicationDriver - 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
INFO  09:00:48.303 [Thread-24] ApplicationDriver - 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
INFO  09:00:48.304 [Thread-24] ApplicationDriver - 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
INFO  09:00:48.304 [Thread-24] ApplicationDriver - 	at java.lang.reflect.Method.invoke(Method.java:483)
INFO  09:00:48.304 [Thread-24] ApplicationDriver - 	at org.apache.spark.deploy.SparkSubmit$.org$apache$spark$deploy$SparkSubmit$$runMain(SparkSubmit.scala:672)
INFO  09:00:48.304 [Thread-24] ApplicationDriver - 	at org.apache.spark.deploy.SparkSubmit$.doRunMain$1(SparkSubmit.scala:180)
INFO  09:00:48.304 [Thread-24] ApplicationDriver - 	at org.apache.spark.deploy.SparkSubmit$.submit(SparkSubmit.scala:205)
INFO  09:00:48.304 [Thread-24] ApplicationDriver - 	at org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:120)
INFO  09:00:48.304 [Thread-24] ApplicationDriver - 	at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala)
INFO  09:00:48.311 [Thread-24] ApplicationDriver - 16/03/30 09:00:48 INFO ShutdownHookManager: Shutdown hook called
INFO  09:00:48.312 [Thread-24] ApplicationDriver - 16/03/30 09:00:48 INFO ShutdownHookManager: Deleting directory C:\Users\kaspers\AppData\Local\Temp\spark-ad111e6e-1941-44c2-8b55-74ebd08eb869
```

Looked up the log files on the Hadoop server. Here's the relevant part:

```
java.lang.IllegalArgumentException: Illegal character in authority at index 7: hdfs://{Bigdatavm - direct}/weblog/weblog_2016_03_28.txt
	at java.net.URI.create(URI.java:852)
	at org.datacleaner.spark.SparkConfigurationReaderInterceptor.createResource(SparkConfigurationReaderInterceptor.java:76)
	at org.datacleaner.configuration.JaxbConfigurationReader.createDatastore(JaxbConfigurationReader.java:1296)
	at org.datacleaner.configuration.JaxbConfigurationReader.createDatastoreCatalog(JaxbConfigurationReader.java:735)
	at org.datacleaner.configuration.JaxbConfigurationReader.create(JaxbConfigurationReader.java:326)
	at org.datacleaner.configuration.JaxbConfigurationReader.create(JaxbConfigurationReader.java:246)
	at org.datacleaner.configuration.JaxbConfigurationReader.read(JaxbConfigurationReader.java:217)
	at org.datacleaner.spark.SparkJobContext.getConfiguration(SparkJobContext.java:126)
	at org.datacleaner.spark.SparkJobContext.getAnalysisJobBuilder(SparkJobContext.java:153)
	at org.datacleaner.spark.SparkJobContext.getAnalysisJob(SparkJobContext.java:145)
	at org.datacleaner.spark.SparkAnalysisRunner.run(SparkAnalysisRunner.java:87)
	at org.datacleaner.spark.Main.main(Main.java:76)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.spark.deploy.yarn.ApplicationMaster$$anon$2.run(ApplicationMaster.scala:525)
Caused by: java.net.URISyntaxException: Illegal character in authority at index 7: hdfs://{Bigdatavm - direct}/weblog/weblog_2016_03_28.txt
	at java.net.URI$Parser.fail(URI.java:2848)
	at java.net.URI$Parser.parseAuthority(URI.java:3186)
	at java.net.URI$Parser.parseHierarchical(URI.java:3097)
	at java.net.URI$Parser.parse(URI.java:3053)
	at java.net.URI.<init>(URI.java:588)
	at java.net.URI.create(URI.java:850)
	... 16 more
```

--- a/engine/core/src/main/java/org/datacleaner/util/convert/HadoopResourceBuilder.java
+++ b/engine/core/src/main/java/org/datacleaner/util/convert/HadoopResourceBuilder.java
@@ -45,7 +45,7 @@ public class HadoopResourceBuilder {
      * <li>Group 3: The path (example '/foo/bar.txt')</li>
      * </ul>
      */
-    public static final Pattern RESOURCE_SCHEME_PATTERN = Pattern.compile("([\\w\\+\\-\\.]+)://\\{([\\w\\.]*)\\}(.*)");
+    public static final Pattern RESOURCE_SCHEME_PATTERN = Pattern.compile("([\\w\\+\\-\\.]+)://\\{([\\w\\.\\W\\s]*)\\}(.*)");
 
     public HadoopResourceBuilder(ServerInformationCatalog catalog, String templatedUri) {
         final Matcher matcher = RESOURCE_SCHEME_PATTERN.matcher(templatedUri);

--- a/engine/core/src/test/java/org/datacleaner/util/convert/HadoopResourceBuilderTest.java
+++ b/engine/core/src/test/java/org/datacleaner/util/convert/HadoopResourceBuilderTest.java
@@ -37,6 +37,16 @@ public class HadoopResourceBuilderTest {
     }
 
     @Test
+    public void testPatternGroupsWithSpaces() throws Exception {
+        final Matcher matcher = HadoopResourceBuilder.RESOURCE_SCHEME_PATTERN.matcher(
+                "hdfs://{my /-+\\|& server}/foo/bar.txt");
+        assertTrue(matcher.find());
+        assertEquals("hdfs", matcher.group(1));
+        assertEquals("my /-+\\|& server", matcher.group(2));
+        assertEquals("/foo/bar.txt", matcher.group(3));
+    }
+
+    @Test
     public void testPatternGroupsNoScheme() throws Exception {
         final Matcher matcher = HadoopResourceBuilder.RESOURCE_SCHEME_PATTERN.matcher("/foo/bar.txt");
         assertFalse(matcher.find());

